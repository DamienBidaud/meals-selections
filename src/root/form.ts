import { atom } from "recoil";

type Form = {
  meal: string;
  guests: number;
  restaurant: string;
  selectedMeals: Array<Meal>;
};

export type Meal = { meal: string; serving: number };

export const formState = atom<Form>({
  key: "form",
  default: {
    meal: "",
    guests: 1,
    restaurant: "",
    selectedMeals: [
      {
        meal: "",
        serving: 1,
      },
    ],
  },
});
