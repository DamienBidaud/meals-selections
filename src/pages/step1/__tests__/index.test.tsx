import { fireEvent, render, RenderResult, screen } from "@testing-library/react";
import { useEffect } from "react";
import { RecoilRoot, useRecoilValue } from "recoil";
import Step1 from "../index";
import { formState } from "../../../root/form";
import { createBrowserRouter, RouterProvider } from "react-router-dom";
import * as routerFunction from "react-router";
import userEvent from "@testing-library/user-event";

//@ts-ignore
export const RecoilObserver = ({ node, onChange }) => {
  const value = useRecoilValue(node);
  useEffect(() => onChange(value), [onChange, value]);
  return null;
};

const mockNavigate = jest.fn();
const mockUseNavigate = jest.spyOn(routerFunction, "useNavigate");

const router = createBrowserRouter([
  {
    path: "/",
    element: <Step1 />,
  },
]);

const onChange = jest.fn();
describe("Step 1", () => {
  let step1: RenderResult;
  beforeEach(() => {
    step1 = render(
      <RecoilRoot>
        <RouterProvider router={router} />
        <RecoilObserver node={formState} onChange={onChange} />
      </RecoilRoot>
    );
  });
  it("Should match the snapshot", () => {
    const { asFragment } = step1;
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should not navagitate when no restaurant have been selected", () => {
    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    mockUseNavigate.mockReturnValue(mockNavigate);

    userEvent.click(screen.getByText("Next"));
    expect(mockNavigate).not.toHaveBeenCalled();
    expect(alertMock).toHaveBeenCalledTimes(1);
  });

  it("Should navigate when a restaurant have been selected", () => {
    const alertMock = jest.spyOn(window, "alert").mockImplementation();
    mockUseNavigate.mockReturnValue(mockNavigate);
    userEvent.selectOptions(screen.getByRole("combobox"), "dinner");
    fireEvent.change(screen.getByRole("spinbutton"), { target: { value: 3 }}); // works perfect

    userEvent.click(screen.getByText("Next"));
    expect(onChange).toHaveBeenCalledWith({
      guests: 3,
      meal: "dinner",
      restaurant: "",
      selectedMeals: [
        {
          meal: "",
          serving: 1,
        },
      ],
    });
    expect(mockNavigate).toHaveBeenCalled();
    expect(alertMock).toHaveBeenCalledTimes(0);
  });
});
