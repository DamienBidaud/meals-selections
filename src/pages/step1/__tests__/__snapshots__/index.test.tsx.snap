// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step 1 Should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-dkrFOg jwbFnv"
  >
    <h1
      class="sc-bcXHqe iokXfP"
    >
      Step 1
    </h1>
    <div>
      <label
        for="meal"
      >
        Select a meal
      </label>
      <select
        id="meal"
        required=""
      >
        <option
          disabled=""
          selected=""
          value=""
        >
          ----
        </option>
        <option
          value="breakfast"
        >
          breakfast
        </option>
        <option
          value="lunch"
        >
          lunch
        </option>
        <option
          value="dinner"
        >
          dinner
        </option>
      </select>
    </div>
    <div>
      <label
        for="guest"
      >
        Number of guest
      </label>
      <input
        id="guest"
        max="10"
        min="1"
        required=""
        type="number"
        value="1"
      />
    </div>
    <button
      class="sc-gswNZR sc-hLBbgP bzVgbG bHsA-dW"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;
